colour/volume/macadam_limits.py

```diff
@@ -108,3 +109,187 @@ def is_within_macadam_limits(xyY, illuminant, tolerance=None):
+        Yxy_Narrowband_Spectra = colour.XYZ_to_xy(
+        colour.macadam_limits(n/20, illuminant) / 100)
+        plot_Narrowband_Spectra (Yxy_Narrowband_Spectra)
+
+    plt.show()
+    """
+    target_bright = target_brightness
+    if target_bright > 1 or target_bright < 0:
+        raise TypeError('brightness of function macadam_limits( )'
+                        'has to be between 0 and 1')
+    standard_cfms = MSDS_CMFS['CIE 1931 2 Degree Standard Observer']
+    X_cie31 = standard_cfms.values[..., 0]
+    Y_cie31 = standard_cfms.values[..., 1]
+    Z_cie31 = standard_cfms.values[..., 2]
+    try:
+        illuminant.interpolator
+    except AttributeError:
+        illuminant = SDS_ILLUMINANTS['E']
+
+
+# If there is no illuminant or it has the wrong form,
+# an illuminant choosen with no influence
+# If the illuminanats do not match the format of the Standard Observer,
+# they have to be adaptet
+    illuminant.extrapolate(SpectralShape(360, 830))
+    illuminant.interpolate(SpectralShape(360, 830, 1))
+    # The cie31 cmfs are convolved with the given illuminant
+    X_illuminated = X_cie31 * illuminant.values
+    Y_illuminated = Y_cie31 * illuminant.values
+    Z_illuminated = Z_cie31 * illuminant.values
+    # Generate empty output-array
+    out_limits = np.zeros_like(standard_cfms.values)
+    # This Array has 471 entries for whavelenghts from 360 nm to 830 nm
+    opti_colour = np.zeros_like(Y_illuminated)
+    # The array of optimal colours has the same dimensions like Y_illuminated
+    # and all entries are initialy set to zero
+    middle_opti_colour = 235
+    # is a constant and not be changed. At 595nm (360 + 235)
+    # in the middle of the center_opti_colour-array
+    # be aware that counting in array-positions starts at zero
+    # The first optimum color has its center initialy at zero
+    maximum_brightness = np.sum(Y_illuminated)
+
+    # "integral" over Y_illuminated
+
+    def optimum_colour(width, center):
+        opti_colour = np.zeros(471)
+        # creates array of 471 zeros and ones which represents optimum-colours
+        # All values of the opti_colour-array are intialy set to zero
+        half_width = width
+        center_opti_colour = center
+        middle_opti_colour = 235
+        opti_colour[middle_opti_colour - half_width:middle_opti_colour +
+                    half_width + 1] = 1
+        # we start the construction of the optimum color
+        # at the center of the opti_colour-array
+        opti_colour = np.roll(opti_colour,
+                              center_opti_colour - middle_opti_colour)
+        # the optimum colour is rolled to the right whavelenght
+        return opti_colour
+
+    def bright_opti_colour(width, center, lightsource):
+        brightness = np.sum(
+            optimum_colour(width, center) * lightsource) / maximum_brightness
+        return brightness
+
+    step_size = np.array([64, 32, 16, 8, 4, 2, 1])
+    for whavelength in range(0, 471):
```

Comment:
```suggestion
def macadam_limits(target_brightness, illuminant=()):
    """
    wavelength reaches from 360 to 830 nm, in within the program it is
    handled as 0 to 470. Beyond the references this program is very fast,
    because the possible optimums are not simply tested step by step but
    more effectively targeted by steps of power of two. The wavelengths
    left and right of a rough optimum are fited by a rule of proportion,
    so that the desired brightness will be exactly reached.

    Parameters
    ----------
    target_brightness : floating point
        brightness has to be between 0 and 1

    illuminant: object
        illuminant must be out of colorimetry.MSDS_CMFS['XXX']
        If there is no illuminant or it has the wrong form,
        the illuminant SDS_ILLUMINANTS['E']
        is choosen wich has no influence to the calculations,
        because it is an equal-energy-spectrum

    if necessary a third parameter for the
    colour-matching funciton could easily be implemented

    Returns
    -------
    an array of CIE -X,Y,Z - Triples for every single wavelength
    in single nm - Steps in the range from 360 to 830 nm

    References
    ----------
    -   cite: Wyszecki, G., & Stiles, W. S. (2000).
        In Color Science: Concepts and Methods,
        Quantitative Data and Formulae (pp. 181–184). Wiley.
        ISBN:978-0-471-39918-6
    -   cite: Francisco Martínez-Verdú, Esther Perales,
        Elisabet Chorro, Dolores de Fez,
        Valentín Viqueira, and Eduardo Gilabert, "Computation and
        visualization of the MacAdam limits
        for any lightness, hue angle, and light source," J.
        Opt. Soc. Am. A 24, 1501-1515 (2007)
    -   cite: Kenichiro Masaoka. In OPTICS LETTERS, June 15, 2010
        / Vol. 35, No. 1 (pp. 2031 - 2033)

    Example
    --------
    from matplotlib import pyplot as plt
    import numpy as np
    import math
    fig = plt.figure(figsize=(7,7))
    ax = fig.add_axes([0,0,1,1])
    illuminant = colour.SDS_ILLUMINANTS['D65']

    def plot_Narrowband_Spectra (Yxy_Narrowband_Spectra):
        FirstColumn = 0
        SecondColumn = 1

        x = Yxy_Narrowband_Spectra[...,FirstColumn]
        y = Yxy_Narrowband_Spectra[...,SecondColumn]
        ax.plot(x,y,'orange',label='Spectrum Loci')

        x = [Yxy_Narrowband_Spectra[-1][FirstColumn],
        Yxy_Narrowband_Spectra[0][FirstColumn]]
        y = [Yxy_Narrowband_Spectra[-1][SecondColumn],
        Yxy_Narrowband_Spectra[0][SecondColumn]]
        ax.plot(x,y,'purple',label='Purple Boundary')
        return()

    for n in range(1, 20):
        Yxy_Narrowband_Spectra = colour.XYZ_to_xy(
        colour.macadam_limits(n/20, illuminant) / 100)
        plot_Narrowband_Spectra (Yxy_Narrowband_Spectra)

    plt.show()
    """
    target_bright = target_brightness
    if target_bright > 1 or target_bright < 0:
        raise TypeError('brightness of function macadam_limits( )'
                        'has to be between 0 and 1')
    standard_cfms = MSDS_CMFS['CIE 1931 2 Degree Standard Observer']
    X_cie31 = standard_cfms.values[..., 0]
    Y_cie31 = standard_cfms.values[..., 1]
    Z_cie31 = standard_cfms.values[..., 2]
    try:
        illuminant.interpolator
    except AttributeError:
        illuminant = SDS_ILLUMINANTS['E']


# If there is no illuminant or it has the wrong form,
# an illuminant choosen with no influence
# If the illuminanats do not match the format of the Standard Observer,
# they have to be adapted
    illuminant.extrapolate(SpectralShape(360, 830))
    illuminant.interpolate(SpectralShape(360, 830, 1))
    # The cie31 cmfs are convolved with the given illuminant
    X_illuminated = X_cie31 * illuminant.values
    Y_illuminated = Y_cie31 * illuminant.values
    Z_illuminated = Z_cie31 * illuminant.values
    # Generate empty output-array
    out_limits = np.zeros_like(standard_cfms.values)
    # This Array has 471 entries for wavelengths from 360 nm to 830 nm
    opti_colour = np.zeros_like(Y_illuminated)
    # The array of optimal colours has the same dimensions like Y_illuminated
    # and all entries are initialy set to zero
    middle_opti_colour = 235
    # is a constant and not be changed. At 595nm (360 + 235)
    # in the middle of the center_opti_colour-array
    # be aware that counting in array-positions starts at zero
    # The first optimum color has its center initialy at zero
    maximum_brightness = np.sum(Y_illuminated)

    # "integral" over Y_illuminated

    def optimum_colour(width, center):
        opti_colour = np.zeros(471)
        # creates array of 471 zeros and ones which represents optimum-colours
        # All values of the opti_colour-array are intialy set to zero
        half_width = width
        center_opti_colour = center
        middle_opti_colour = 235
        opti_colour[middle_opti_colour - half_width:middle_opti_colour +
                    half_width + 1] = 1
        # we start the construction of the optimum color
        # at the center of the opti_colour-array
        opti_colour = np.roll(opti_colour,
                              center_opti_colour - middle_opti_colour)
        # the optimum colour is rolled to the right wavelength
        return opti_colour

    def bright_opti_colour(width, center, lightsource):
        brightness = np.sum(
            optimum_colour(width, center) * lightsource) / maximum_brightness
        return brightness

    step_size = np.array([64, 32, 16, 8, 4, 2, 1])
    for wavelength in range(0, 471):
```

Reply:
Sorry, I am not a native-speaker. It's a bit embarrassing, to molest You in teaching me school work. I nevermore will misspell wavelength, adapted and program. I'll write each word to my exercise book 100 times ;-)

But really thank You for helping me

Reply:
Please, don't be embarrassed! *I'm* embarrassed, I only speak the one language!  I thought I was just making a tiny suggestion, but then I came to realize that the small change I was suggesting led to more and more changes, and now I've drawn undue attention to your spelling.  I'm sorry for spamming the review / comments here with so much noise. I hope I haven't made you self-conscious -- your contributions are greatly appreciated, and your English is pretty damn fluent!

(Also, Thomas is almost definitely going to end up correcting my English as well : ) )

Reply:
I am always happy if I can learn, even spelling. And I am enthusiastic, that here on github everyone can contribute his expertise to get at the and a excelent result by real teamwork, which no one of us could have done on his own.